In [1]:
from google.colab import files
uploaded=files.upload()

Saving rainfall_area-wt_India_1901-2015 (1).csv to rainfall_area-wt_India_1901-2015 (1).csv


In [2]:
import pandas as pd

rainfall_data = pd.read_csv("/content/rainfall_area-wt_India_1901-2015 (1).csv")

print(rainfall_data)

    REGION  YEAR   JAN   FEB   MAR   APR   MAY    JUN    JUL    AUG    SEP  \
0    INDIA  1901  34.7  37.7  18.0  39.3  50.8  113.4  242.2  272.9  124.4   
1    INDIA  1902   7.4   4.3  19.0  43.5  48.3  108.8  284.0  199.7  201.5   
2    INDIA  1903  17.0   8.3  31.3  17.1  59.5  118.3  297.0  270.4  199.1   
3    INDIA  1904  14.4   9.6  31.8  33.1  72.4  164.8  261.0  206.4  129.6   
4    INDIA  1905  25.3  20.9  42.7  33.7  55.7   93.3  252.8  200.8  178.4   
..     ...   ...   ...   ...   ...   ...   ...    ...    ...    ...    ...   
110  INDIA  2011   7.7  26.3  21.4  41.0  51.6  182.8  243.0  284.6  190.5   
111  INDIA  2012  28.5  10.8  10.6  48.5  32.1  125.0  258.1  263.9  197.7   
112  INDIA  2013  10.0  36.9  14.5  29.4  56.5  217.9  305.4  248.4  148.4   
113  INDIA  2014  17.3  25.9  32.6  20.2  72.8   95.5  258.8  237.9  187.9   
114  INDIA  2015  17.4  21.0  62.0  69.4  53.8  192.8  242.4  205.2  131.8   

       OCT   NOV   DEC  ANNUAL  Jan-Feb  Mar-May  Jun-Sep  Oct-

In [3]:
rainfall_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 115 entries, 0 to 114
Data columns (total 19 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   REGION   115 non-null    object 
 1   YEAR     115 non-null    int64  
 2   JAN      115 non-null    float64
 3   FEB      115 non-null    float64
 4   MAR      115 non-null    float64
 5   APR      115 non-null    float64
 6   MAY      115 non-null    float64
 7   JUN      115 non-null    float64
 8   JUL      115 non-null    float64
 9   AUG      115 non-null    float64
 10  SEP      115 non-null    float64
 11  OCT      115 non-null    float64
 12  NOV      115 non-null    float64
 13  DEC      115 non-null    float64
 14  ANNUAL   115 non-null    float64
 15  Jan-Feb  115 non-null    float64
 16  Mar-May  115 non-null    float64
 17  Jun-Sep  115 non-null    float64
 18  Oct-Dec  115 non-null    float64
dtypes: float64(17), int64(1), object(1)
memory usage: 17.2+ KB


In [4]:
rainfall_data.columns

Index(['REGION', 'YEAR', 'JAN', 'FEB', 'MAR', 'APR', 'MAY', 'JUN', 'JUL',
       'AUG', 'SEP', 'OCT', 'NOV', 'DEC', 'ANNUAL', 'Jan-Feb', 'Mar-May',
       'Jun-Sep', 'Oct-Dec'],
      dtype='object')

In [5]:
rainfall_data.isnull().sum()

REGION     0
YEAR       0
JAN        0
FEB        0
MAR        0
APR        0
MAY        0
JUN        0
JUL        0
AUG        0
SEP        0
OCT        0
NOV        0
DEC        0
ANNUAL     0
Jan-Feb    0
Mar-May    0
Jun-Sep    0
Oct-Dec    0
dtype: int64

In [6]:
rainfall_data.describe()

YEAR         JAN         FEB         MAR         APR  \
count   115.000000  115.000000  115.000000  115.000000  115.000000   
mean   1958.000000   19.759130   23.434783   28.254783   38.241739   
std      33.341666    9.992628   11.512739   12.286408   10.353040   
min    1901.000000    2.700000    2.700000    7.200000   16.100000   
25%    1929.500000   13.000000   13.300000   19.750000   31.600000   
50%    1958.000000   17.800000   22.500000   25.500000   37.400000   
75%    1986.500000   24.850000   30.300000   34.400000   43.850000   
max    2015.000000   58.500000   53.800000   63.300000   69.400000   

              MAY         JUN         JUL         AUG         SEP         OCT  \
count  115.000000  115.000000  115.000000  115.000000  115.000000  115.000000   
mean    62.193913  168.360000  291.022609  258.400870  172.473043   75.701739   
std     15.673378   35.569654   41.161390   34.975419   36.641234   28.268152   
min     32.100000   86.500000  138.900000  191.700000   96.900000   20.000000   
25%     51.600000  144.050000  267.350000  233.950000  144.850000   55.600000   
50%     59.500000  165.600000  295.800000  259.300000  173.100000   69.200000   
75%     71.200000  192.050000  318.650000  287.950000  198.300000   92.850000   
max    114.500000  275.500000  383.400000  335.500000  281.000000  158.800000   

              NOV         DEC       ANNUAL     Jan-Feb     Mar-May  \
count  115.000000  115.000000   115.000000  115.000000  115.000000   
mean    29.205217   14.980000  1182.034783   43.189565  128.694783   
std     16.101056    8.788761   110.686214   14.476335   22.895134   
min      3.600000    1.600000   920.800000   11.700000   84.500000   
25%     17.300000    9.600000  1102.400000   33.800000  112.350000   
50%     26.100000   14.100000  1190.500000   41.300000  125.100000   
75%     39.650000   19.000000  1243.550000   51.400000  139.650000   
max     74.200000   54.400000  1480.300000   86.300000  209.700000   

           Jun-Sep     Oct-Dec  
count   115.000000  115.000000  
mean    890.260870  119.882609  
std      89.179055   32.467126  
min     679.500000   52.700000  
25%     823.550000   97.050000  
50%     897.800000  116.200000  
75%     959.650000  142.150000  
max    1094.500000  207.500000

## EDA Analysis

In [7]:
## Analyzing Trends in Annual Rainfall
import plotly.graph_objects as go ## creating advanced interactive visualizations
import plotly.express as px ## Provides simple methods for data visualization

# analyze trends in annual rainfall over time
annual_rainfall = rainfall_data[['YEAR', 'ANNUAL']] ##  Extracts two columns: YEAR and ANNUAL

## Create a Line Chart for Annual Rainfall Trends
fig_annual = go.Figure()
## Initializes an empty Plotly figure object to store the plot.
fig_annual.add_trace(go.Scatter(
    x=annual_rainfall['YEAR'],
    y=annual_rainfall['ANNUAL'],
    mode='lines',
    name='Annual Rainfall',
    line=dict(color='blue', width=2),
    opacity=0.7
))
fig_annual.add_trace(go.Scatter(
    x=annual_rainfall['YEAR'],
    y=[annual_rainfall['ANNUAL'].mean()] * len(annual_rainfall),
    mode='lines',
    name='Mean Rainfall',
    line=dict(color='red', dash='dash')
))
fig_annual.update_layout(
    title='Trend in Annual Rainfall in India (1901-2015)',
    xaxis_title='Year',
    yaxis_title='Rainfall (mm)',
    template='plotly_white',
    legend=dict(title="Legend"),
    height=500
)
fig_annual.show()

# identify months with the highest and lowest rainfall on average
monthly_columns = ['JAN', 'FEB', 'MAR', 'APR', 'MAY', 'JUN', 'JUL', 'AUG', 'SEP', 'OCT', 'NOV', 'DEC']
monthly_avg = rainfall_data[monthly_columns].mean()

highest_rainfall_month = monthly_avg.idxmax()
lowest_rainfall_month = monthly_avg.idxmin()

## Create a Bar Chart for Average Monthly Rainfall

fig_monthly = px.bar(
    x=monthly_avg.index,
    y=monthly_avg.values,
    labels={'x': 'Month', 'y': 'Rainfall (mm)'},
    title='Average Monthly Rainfall in India (1901-2015)',
    text=monthly_avg.values
)
fig_monthly.add_hline(
    y=monthly_avg.mean(),
    line_dash="dash",
    line_color="red",
    annotation_text="Mean Rainfall",
    annotation_position="top right"
)
fig_monthly.update_traces(marker_color='skyblue', marker_line_color='black', marker_line_width=1)
fig_monthly.update_layout(template='plotly_white', height=500)
fig_monthly.show()

# seasonal rainfall distribution
seasonal_columns = ['Jan-Feb', 'Mar-May', 'Jun-Sep', 'Oct-Dec']
seasonal_avg = rainfall_data[seasonal_columns].mean()

## Draw a Bar plot
fig_seasonal = px.bar(
    x=seasonal_avg.index,
    y=seasonal_avg.values,
    labels={'x': 'Season', 'y': 'Rainfall (mm)'},
    title='Seasonal Rainfall Distribution in India (1901-2015)',
    text=seasonal_avg.values,
    color=seasonal_avg.values,
    color_continuous_scale=['gold', 'skyblue', 'green', 'orange']
)

fig_seasonal.update_traces(marker_line_color='black', marker_line_width=1)
fig_seasonal.update_layout(
    template='plotly_white',
    height=500,
    coloraxis_colorbar=dict(title='mm')
)
fig_seasonal.update_layout(template='plotly_white', height=500)
fig_seasonal.show()

## Observation of 1 st plot

- The graph shows how India's annual rainfall changes each year. The blue line represents yearly rainfall, which goes up and down without a clear increasing or decreasing trend. The red dashed line shows the average rainfall over time. Some years had very high or very low rainfall, indicating extreme weather events.

## Observation of 2 nd plot

- This bar chart illustrates a highly uneven distribution of rainfall across months, with July and August receiving the highest average rainfall. The red dashed line represents the mean monthly rainfall, showing that most months receive rainfall below the average, except during the monsoon months (June to September).

## Observation of 3 rd plot

- The seasonal distribution highlights the dominance of the monsoon season (June to September), which contributes the bulk of annual rainfall (around 890 mm). In contrast, the other seasons (January-February, March-May, and October-December) contribute significantly less to the annual total, which emphasizes the critical role of the monsoon.

## Assessing the Impact of Climate Change in the Rainfall Trends in India

In [8]:
# calculating rolling averages to assess climate change impact
rainfall_data['10-Year Rolling Avg'] = rainfall_data['ANNUAL'].rolling(window=10).mean()

fig_climate_change = go.Figure()

fig_climate_change.add_trace(go.Scatter(
    x=rainfall_data['YEAR'],
    y=rainfall_data['ANNUAL'],
    mode='lines',
    name='Annual Rainfall',
    line=dict(color='blue', width=2),
    opacity=0.6
))

fig_climate_change.add_trace(go.Scatter(
    x=rainfall_data['YEAR'],
    y=rainfall_data['10-Year Rolling Avg'],
    mode='lines',
    name='10-Year Rolling Avg',
    line=dict(color='red', width=3)
))

fig_climate_change.update_layout(
    title='Impact of Climate Change on Rainfall Patterns (1901-2015)',
    xaxis_title='Year',
    yaxis_title='Rainfall (mm)',
    template='plotly_white',
    legend=dict(title="Legend"),
    height=500
)

fig_climate_change.show()

## Identifying Drought and Extreme Rainfall Years



In [9]:
from scipy.stats import pearsonr

# identifying drought and extreme rainfall years
mean_rainfall = rainfall_data['ANNUAL'].mean()
std_dev_rainfall = rainfall_data['ANNUAL'].std()

drought_years = rainfall_data[rainfall_data['ANNUAL'] < (mean_rainfall - 1.5 * std_dev_rainfall)]
extreme_rainfall_years = rainfall_data[rainfall_data['ANNUAL'] > (mean_rainfall + 1.5 * std_dev_rainfall)]

# correlating seasonal rainfall with annual rainfall totals
seasonal_columns = ['Jan-Feb', 'Mar-May', 'Jun-Sep', 'Oct-Dec']
seasonal_correlations = {
    season: pearsonr(rainfall_data[season], rainfall_data['ANNUAL'])[0] for season in seasonal_columns
}

# displaying results for drought/extreme years and correlations
drought_years_summary = drought_years[['YEAR', 'ANNUAL']].reset_index(drop=True)
extreme_rainfall_years_summary = extreme_rainfall_years[['YEAR', 'ANNUAL']].reset_index(drop=True)
seasonal_correlations_summary = pd.DataFrame.from_dict(seasonal_correlations, orient='index', columns=['Correlation'])

drought_years_summary, extreme_rainfall_years_summary, seasonal_correlations_summary

(   YEAR  ANNUAL
 0  1905   975.3
 1  1965   938.4
 2  1972   948.5
 3  2002   920.8
 4  2009   959.3,
    YEAR  ANNUAL
 0  1917  1480.3
 1  1933  1393.5
 2  1956  1386.2
 3  1959  1382.1
 4  1961  1403.0
 5  1988  1351.0
 6  1990  1400.6,
          Correlation
 Jan-Feb     0.228913
 Mar-May     0.313057
 Jun-Sep     0.930027
 Oct-Dec     0.531648)

## Detecting Anomalies in the Rainfall Trends in India

In [10]:
from sklearn.ensemble import IsolationForest

# detect anomalous rainfall years based on annual data
isolation_forest = IsolationForest(contamination=0.05, random_state=42)
rainfall_data['Annual_Anomaly'] = isolation_forest.fit_predict(rainfall_data[['ANNUAL']])

# identify anomalies in annual rainfall
annual_anomalies = rainfall_data[rainfall_data['Annual_Anomaly'] == -1]

# detect anomalous months based on monthly data
monthly_data = rainfall_data[['JAN', 'FEB', 'MAR', 'APR', 'MAY', 'JUN', 'JUL', 'AUG', 'SEP', 'OCT', 'NOV', 'DEC']]
monthly_anomalies = isolation_forest.fit_predict(monthly_data)

# add anomaly detection results for months
rainfall_data['Monthly_Anomaly'] = monthly_anomalies
monthly_anomalies_df = rainfall_data[rainfall_data['Monthly_Anomaly'] == -1][['YEAR'] + monthly_columns]

fig_annual_anomalies = go.Figure()

fig_annual_anomalies.add_trace(go.Scatter(
    x=rainfall_data['YEAR'],
    y=rainfall_data['ANNUAL'],
    mode='lines',
    name='Annual Rainfall',
    line=dict(color='blue', width=2),
    opacity=0.6
))

fig_annual_anomalies.add_trace(go.Scatter(
    x=annual_anomalies['YEAR'],
    y=annual_anomalies['ANNUAL'],
    mode='markers',
    name='Anomalous Years',
    marker=dict(color='red', size=8, symbol='circle')
))

fig_annual_anomalies.add_hline(
    y=rainfall_data['ANNUAL'].mean(),
    line_dash='dash',
    line_color='green',
    annotation_text='Mean Rainfall',
    annotation_position='bottom right'
)

fig_annual_anomalies.update_layout(
    title='Annual Rainfall Anomalies in India (1901-2015)',
    xaxis_title='Year',
    yaxis_title='Rainfall (mm)',
    template='plotly_white',
    legend=dict(title="Legend"),
    height=500
)

fig_annual_anomalies.show()

##  identify anomalies in monthly rainfall:

In [11]:
# preparing data for monthly anomalies
monthly_anomalies = []
for column in monthly_columns:
    for _, row in monthly_anomalies_df.iterrows():
        monthly_anomalies.append({'Year': row['YEAR'], 'Month': column, 'Rainfall': row[column]})

monthly_anomalies_df_long = pd.DataFrame(monthly_anomalies)

fig_monthly_anomalies = px.line(
    rainfall_data,
    x='YEAR',
    y=monthly_columns,
    labels={'YEAR': 'Year', 'value': 'Rainfall (mm)', 'variable': 'Month'},
    title='Monthly Rainfall Anomalies in India (1901-2015)',
    color_discrete_sequence=px.colors.qualitative.Set3
)

fig_monthly_anomalies.add_trace(go.Scatter(
    x=monthly_anomalies_df_long['Year'],
    y=monthly_anomalies_df_long['Rainfall'],
    mode='markers',
    name='Anomalous Months',
    marker=dict(color='red', size=5, symbol='circle')
))

fig_monthly_anomalies.update_layout(
    template='plotly_white',
    legend=dict(title="Legend"),
    height=500
)

fig_monthly_anomalies.show()

## Correlating Seasonal Rainfall with Annual Totals

In [12]:
# correlation analysis between monsoon (Jun-Sep) rainfall and other seasons
seasonal_columns = ['Jan-Feb', 'Mar-May', 'Jun-Sep', 'Oct-Dec']
monsoon_column = 'Jun-Sep'
relationships = {}

for season in seasonal_columns:
    if season != monsoon_column:
        corr, _ = pearsonr(rainfall_data[monsoon_column], rainfall_data[season])
        relationships[season] = corr

correlation_data = pd.DataFrame({
    'Season': list(relationships.keys()),
    'Correlation Coefficient': list(relationships.values())
})

fig = px.bar(
    correlation_data,
    x='Season',
    y='Correlation Coefficient',
    title='Correlation Between Monsoon (Jun-Sep) Rainfall and Other Seasons',
    labels={'Season': 'Season', 'Correlation Coefficient': 'Correlation Coefficient'},
    text='Correlation Coefficient',
    color='Correlation Coefficient',
    color_continuous_scale='Blues'
)

fig.add_hline(
    y=0,
    line_dash="dash",
    line_color="red",
    annotation_text="No Correlation",
    annotation_position="bottom left"
)

fig.update_traces(marker_line_color='black', marker_line_width=1, texttemplate='%{text:.2f}')
fig.update_layout(
    template='plotly_white',
    height=500
)

fig.show()

## Grouping Years Based on Rainfall Patterns

In [13]:
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler

# prepare data for clustering
rainfall_features = rainfall_data[['Jan-Feb', 'Mar-May', 'Jun-Sep', 'Oct-Dec', 'ANNUAL']]
scaler = StandardScaler()
scaled_features = scaler.fit_transform(rainfall_features)

# perform k-means clustering
kmeans = KMeans(n_clusters=3, random_state=42)
rainfall_data['Rainfall_Cluster'] = kmeans.fit_predict(scaled_features)

# map cluster labels to categories (e.g., Dry, Normal, Wet)
cluster_labels = {0: 'Dry', 1: 'Normal', 2: 'Wet'}
rainfall_data['Rainfall_Category'] = rainfall_data['Rainfall_Cluster'].map(cluster_labels)

fig = px.scatter(
    rainfall_data,
    x='YEAR',
    y='ANNUAL',
    color='Rainfall_Category',
    title='Clustering of Years Based on Rainfall Patterns',
    labels={'YEAR': 'Year', 'ANNUAL': 'Annual Rainfall (mm)', 'Rainfall_Category': 'Rainfall Category'},
    color_discrete_sequence=px.colors.qualitative.Set2,
    hover_data={'Rainfall_Cluster': True, 'Rainfall_Category': True}
)

fig.update_layout(
    template='plotly_white',
    legend_title='Rainfall Category',
    height=500
)

fig.show()

## Forecasting Future Rainfall

In [15]:
import pandas as pd
import logging
from prophet import Prophet
from prophet.plot import plot_plotly

# Suppress unnecessary logging
logging.getLogger('cmdstanpy').setLevel(logging.ERROR)

# Convert YEAR column to datetime format
rainfall_data['DATE'] = pd.to_datetime(rainfall_data['YEAR'], format='%Y')

# Create a time series index
annual_rainfall_ts = rainfall_data.set_index('DATE')['ANNUAL']

# Prepare data for Prophet
prophet_data = annual_rainfall_ts.reset_index()
prophet_data.columns = ['ds', 'y']

# Initialize Prophet model with explicit seasonality settings
prophet_model = Prophet(
    yearly_seasonality=True,
    weekly_seasonality=False,  # Disabling weekly seasonality explicitly
    daily_seasonality=False    # Disabling daily seasonality explicitly
)
prophet_model.fit(prophet_data)

# Create a future dataframe for the next 20 years (fixing 'YE' warning)
future = prophet_model.make_future_dataframe(periods=20, freq='YE')
forecast = prophet_model.predict(future)

# Interactive Forecast Plot
fig_forecast = plot_plotly(prophet_model, forecast)
fig_forecast.update_layout(
    title='Annual Rainfall Forecast Using Prophet',
    xaxis_title='Year',
    yaxis_title='Rainfall (mm)',
    template='plotly_white',
    height=500
)

fig_forecast.show()



The analysis of India’s rainfall trends and patterns from 1901 to 2015 reveals significant variability in annual and seasonal rainfall, with the monsoon season (June-September) being the dominant contributor. Anomalous years of extreme drought and wetness highlight the unpredictability of rainfall, while clustering shows a shift towards more dry years in recent decades. Correlations indicate the limited dependency of non-monsoon seasons on monsoon rainfall. A time-series forecast using Prophet suggests a slight declining trend in annual rainfall, which emphasises the need for long-term water resource planning and adaptation to changing climate patterns.

